### ISSUE TITLE
Use proxy settings for downloading

### ISSUE THREAD

**Issue by bekce** [flapdoodle-oss/de.flapdoodle.embed.mongo#271]
Java has preferred system properties for defining proxy settings named `http.proxyHost, http.proxyPort, https.proxyHost, https.proxyPort`. The downloader for the mongod binary package should automatically use these settings if no other setting was provided, by default.

**Comment by ogerardin** [comment]
I have implemented this with the help of proxy-vole https://github.com/MarkusBernhardt/proxy-vole

Concretely: I have a utility class `ProxyHelper `that performs proxy detection and returns a `IProxyFactory `that can be used to configure a `DownloadConfigBuilder`, e.g. as follows:

 ```
class CustomRuntimeConfigBuilder extends de.flapdoodle.embed.mongo.config.RuntimeConfigBuilder {
        /**
         * @return A {@link RuntimeConfigBuilder} set to use a proxy obtained through {@link ProxyHelper}
         */
        public RuntimeConfigBuilder defaultsWithLoggerAndProxy(Command command, Logger logger) {
            super.defaultsWithLogger(command, logger);

            // automatic proxy detecttion
            val proxyFactory = ProxyHelper.detectProxyFactory();

            val downloadConfig = new DownloadConfigBuilder()
                    .defaultsForCommand(command)
                    .proxyFactory(proxyFactory)
                    .progressListener(new Slf4jProgressListener(logger))
                    .build();

            val artifactStore = new ExtractedArtifactStoreBuilder()
                    .defaults(command)
                    .download(downloadConfig)
                    .build();

            artifactStore().overwriteDefault(artifactStore);

            return this;
        }
```

Code for ProxyHelper (imports omitted):
```
@Slf4j
enum ProxyHelper {

    ; //memberless enum

    /**
     * Returns an {@link IProxyFactory} matching the system's proxy settings.
     * We use proxy-vole to detect proxy configuration, see: https://github.com/MarkusBernhardt/proxy-vole
     */
    static IProxyFactory detectProxyFactory() {
        // if the logger is enabled for debug, set an adapter to log debug output of proxy-vole
        if (log.isDebugEnabled()) {
            com.github.markusbernhardt.proxy.util.Logger.setBackend(
                    (clazz, loglevel, msg, params) -> log.debug(MessageFormat.format(msg, params))
            );
        }

        ProxySearch proxySearch = ProxySearch.getDefaultProxySearch();
        ProxySelector proxySelector = proxySearch.getProxySelector();
        if (proxySelector != null) {
            List<Proxy> proxies = proxySelector.select(URI.create("https://mongodb.org"));
            for (Proxy proxy : proxies) {
                try {
                    return getProxyFactory(proxy);
                } catch (Exception e) {
                    log.debug("Failed to convert " + proxy + " to IProxyFactory", e);
                }
            }
        }
        return new NoProxyFactory();
    }

    /**
     * Convert a Java {@link Proxy} object to a {@link IProxyFactory} suitable for
     * passing to {@link DownloadConfigBuilder#proxyFactory}
     *
     * @throws IllegalArgumentException if the proxy type is not supported
     */
    private static IProxyFactory getProxyFactory(Proxy proxy) {
        Proxy.Type proxyType = proxy.type();
        SocketAddress proxyAddress = proxy.address();

        switch (proxyType) {
            case DIRECT:
                return new NoProxyFactory();
            case HTTP: {
                if (proxyAddress == null) {
                    throw new IllegalArgumentException("Proxy address is null");
                }
                if (!(proxyAddress instanceof InetSocketAddress)) {
                    throw new IllegalArgumentException("Unsupported proxy address type: " + proxyAddress.getClass());
                }
                InetSocketAddress inetSocketAddress = (InetSocketAddress) proxyAddress;
                return new HttpProxyFactory(inetSocketAddress.getHostName(), inetSocketAddress.getPort());
            }
            default:
                throw new IllegalArgumentException("Unsupported proxy type: " + proxyType);
        }
    }
}
```

**Comment by roeniss** [comment]
Is this not yet implemented?

**Comment by eidottermihi** [comment]
This seems to be already resolved (tested with `v3.4.8`), system properties like `-Dhttps.proxyHost=my.proxy.com -Dhttps.proxyPort=80` are used for downloading when set.

**Comment by michaelmosmann** [comment]
@eidottermihi yes.. it should:) .. missed this one.. close this.
